test/src/test/java/hudson/model/QueueTest.java

```diff
@@ -1187,4 +1193,111 @@ public void run() {
             execute(new BuildExecution());
         }
     }
+    
+    /**
+     * A simple task, that waits eternally and allows concurrent builds.
+     * 
+     * @author Martin Schroeder
+     */
+    private static class InfiniteWaitTask extends TestTask{
+        InfiniteWaitTask(AtomicInteger cnt) {
+            super(cnt);
+        }
+        
+        @Override
+        public void doRun() {
+            try {
+                Thread.sleep(Integer.MAX_VALUE);
+            } catch (InterruptedException e) { }
+        }
+        
+        @Override
+        public boolean isConcurrentBuild() {
+            //Test must be concurrent, else buildables.size() is always == 1
+            return true;
+        }
+    }
+    
+    /**
+     * This test checks if {@link QueueTaskFuture#cancel(boolean)}
+     * actually works correctly, and cancels the item, instead of the first
+     * random {@link Task} that it finds in the Queue.
+     * <p>
+     * This is, because previous implementations of Jenkins had a design
+     * defect, that caused the wrong task to be cancelled.
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void cancelFutureOfTask() throws Exception {
+        Queue q = r.jenkins.getQueue();
+        
+        //Create an infinitely waiting, concurrent task
+        Task t = new InfiniteWaitTask(new AtomicInteger(0));
+        
+        //Ensure that only a single executor is available
+        r.jenkins.setNumExecutors(1);
+        r.jenkins.reload();
+        
+        //Schedule ONE job that will block the one executor
+        //This is done, so the other jobs will be buildable, not blocked
+        q.schedule2(t, 0);
+        //Wait, for it to be scheduled
+        Thread.sleep(1000);
```

Comment:
Can the test wait for a specific condition to hold, rather than just sleeping and hoping for the best? This looks like it will be flaky.